File: Loom.Providers/OpenAI/OpenAIAdapter.vb
Imports Newtonsoft.Json
Imports Loom.Core.Interfaces
Imports Loom.Core.Models
Imports Loom.Engine
Imports Loom.Providers.Core
Namespace OpenAI
    ''' <summary>
    ''' Provider for OpenAI solutions
    ''' </summary>
    Public Class OpenAIAdapter
        Implements IProviderAdapter

        Private ReadOnly _assembler As IAssembler
        Private ReadOnly _http As New Infrastructure.ProviderHttpClient()
        Private ReadOnly _apiKey As String

        Public Sub New(apiKey As String, assembler As IAssembler)
            _apiKey = apiKey
            _assembler = assembler
        End Sub

        Public ReadOnly Property ProviderName As String Implements IProviderAdapter.ProviderName
            Get
                Return "OpenAI"
            End Get
        End Property

        Public Async Function ExecuteAsync(invocation As ILlmInvocation) As Task(Of LlmResponse) Implements IProviderAdapter.ExecuteAsync

            Dim concreteInvocation = DirectCast(invocation, LlmInvocation)

            Dim inputItems = MappingAssembledInput(concreteInvocation)

            Dim modelId = If(concreteInvocation.Hints.PreferredModel, "gpt-4.1-nano")

            Dim toolList = BuildToolSchema(invocation)

            Dim payload As New Dictionary(Of String, Object) From {
                {"model", modelId},
                {"include", New List(Of String) From {"file_search_call.results"}},
                {"input", inputItems},
                {"tools", toolList},
                {"metadata", New Dictionary(Of String, Object)()},
                {"service_tier", "priority"},
                {"max_output_tokens", invocation.Options.MaxTokens},
                {"temperature", invocation.Options.Temperature}
            }

            Dim headers = New Dictionary(Of String, String) From {
                {"Authorization", $"Bearer {_apiKey}"}
            }


            Dim rawResponse As ModelResponse = Await _http.PostAsync(Of ModelResponse)(headers, "https://api.openai.com/v1/responses", payload)

            Return New LlmResponse With {
                .Content = ExtractResponse(rawResponse),
                .Id = rawResponse.Id,
                .ModelUsed = rawResponse.Model,
                .Role = 2, ' Assistant
                .ToolCalls = ExtractToolInfo(rawResponse),
                .PromptTokens = rawResponse.Usage.Input_tokens,
                .CompletionTokens = rawResponse.Usage.Output_tokens,
                .TotalTokens = rawResponse.Usage.Total_tokens,
                .Metadata = rawResponse.Metadata
            }
        End Function

        Public Function SupportsCapability(capability As String) As Boolean Implements IProviderAdapter.SupportsCapability
            Dim supported = {"vision", "tools", "json_mode"}

            Return supported.Contains(capability.ToLower())
        End Function

        Private Function MappingAssembledInput(concreteInvocation As LlmInvocation) As List(Of Object)
            Dim inputItems = _assembler.AssembleInputItems(concreteInvocation)
            Try
                Return inputItems.Select(Function(x)
                                             If x.Type = "function_call" Then
                                                 Return CType(New With {
                                                                .type = x.Type,
                                                                .name = x.ToolName,
                                                                .arguments = JsonConvert.SerializeObject(x.Arguments),
                                                                .call_id = x.ToolCallId
                                                            }, Object)
                                             ElseIf x.Type = "function_call_output" Then
                                                 Return CType(New With {
                                                                .type = x.Type,
                                                                .call_id = x.ToolCallId,
                                                                .output = x.Content
                                                            }, Object)
                                             Else
                                                 Return CType(New With {
                                                                .role = x.Role,
                                                                .content = x.Content
                                                            }, Object)
                                             End If
                                         End Function).ToList()
            Catch ex As Exception
                Throw New Exception($"[ERROR] OpenAIAdapter.MappingAssembledInput: {ex}")
            End Try

        End Function

        Private Function ExtractResponse(rawResponse As ModelResponse) As String

            Dim risposta_modello As String = ""

            Try

                For Each item In rawResponse.Output
                    If item.Type = "message" Then
                        For Each content In item.Content
                            If content.Type = "output_text" Then
                                risposta_modello = content.Text
                            End If
                        Next
                    End If
                Next
                Return risposta_modello
            Catch ex As Exception
                Throw New Exception($"[ERROR] The response received from the network call doesn't meet the correct parameters: {ex}")
            End Try



        End Function

        Private Function ExtractToolInfo(rawResponse As ModelResponse) As List(Of ToolCallInvocation)

            Dim toolList As New List(Of ToolCallInvocation)

            Try
                For Each item In rawResponse.Output

                    If item.Type = "function_call" OrElse item.Type = "tool_call" Then

                        Dim argsDict As Dictionary(Of String, Object) = Nothing

                        If Not String.IsNullOrEmpty(item.Arguments) Then
                            argsDict = JsonConvert.DeserializeObject(Of Dictionary(Of String, Object))(item.Arguments)
                        Else
                            argsDict = New Dictionary(Of String, Object)
                        End If

                        Dim tool As New ToolCallInvocation With {
                            .Type = "function_call",
                            .CallId = item.Call_id,
                            .ToolName = item.Name,
                            .Arguments = argsDict
                        }

                        toolList.Add(tool)

                    End If

                Next
            Catch ex As Exception
                Throw New Exception($"[ERROR] While parsing tool call: {ex.Message}", ex)
            End Try

            Return toolList

        End Function

        Private Function BuildToolSchema(invocation As ILlmInvocation) As List(Of ProviderToolSchema)

            Dim toolSchemaList As New List(Of ProviderToolSchema)

            Dim builder As New OpenAIToolSchemaBuilder()

            If invocation.Tools.RegisteredTools IsNot Nothing Then
                For Each tool In invocation.Tools.RegisteredTools
                    toolSchemaList.Add(builder.Build(tool))
                Next
            End If
            Return toolSchemaList
        End Function


    End Class
End Namespace

File: implementations/csharp/src/Loom.Providers.OpenAI/OpenAIAdapter.cs
using Loom.Core.Enums;
using Loom.Core.Interfaces;
using Loom.Core.Models;
using Loom.Engine;
using Loom.Engine.Infrastructure;
using Loom.Providers.OpenAI.Schemas;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace Loom.Providers.OpenAI
{
    public class OpenAIAdapter : IProviderAdapter
    {

        private readonly IAssembler _assembler;
        private readonly ProviderHttpClient _http = new ProviderHttpClient();
        private readonly string _apikey;

        public string ProviderName => "OpenAI";

        public OpenAIAdapter(string apikey, IAssembler assembler)
        {
            _apikey = apikey;
            _assembler = assembler;
        }

        /// <summary>
        /// Provider core function:
        /// - Takes the current invocation
        /// - Maps invocation input
        /// - Builds the tool schema
        /// - Assembles the API calls payload + headers
        /// - Makes API calls
        /// - Extracts parameters from the call's response
        /// - Return the response
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns> LlmResponse </returns>
        public async Task<LlmResponse> ExecuteAsync(ILlmInvocation invocation)
        {
            var concreteInvocation = (LlmInvocation)invocation;

            var inputItems = MappingAssembledInput(concreteInvocation);

            // gpt-4.1-nano default
            var modelId = concreteInvocation.Hints?.PreferredModel ?? "gpt-4.1-nano";

            var toolList = BuildToolSchema(invocation);

            var payload = new Dictionary<string, object>
            {
                {"model", modelId },
                { "include", new List<string> { "file_search_call.results" } },
                { "input", inputItems },
                { "tools", toolList },
                { "metadata", new Dictionary<string, object>() },
                { "service_tier", "priority" },
                { "max_output_tokens", invocation.Options.MaxTokens },
                { "temperature", invocation.Options.Temperature }
            };

            var headers = new Dictionary<string, string>
            {
                {"Authorization", $"Bearer {_apikey}"}
            };

            ModelResponse rawResponse = await _http.PostAsync<ModelResponse>(headers, "https://api.openai.com/v1/responses", payload);

            return new LlmResponse
            {
                Content = ExtractResponse(rawResponse),
                Id = rawResponse.Id,
                ModelUsed = rawResponse.Model,
                Role = MessageRole.Assistant,
                ToolCalls = ExtractToolInfo(rawResponse),
                PromptTokens = rawResponse.Usage.Input_tokens,
                CompletionTokens = rawResponse.Usage.Output_tokens,
                TotalTokens = rawResponse.Usage.Total_tokens,
                Metadata = rawResponse.Metadata
            };
        }


        /// <summary>
        /// TODO: Revisiting the function
        /// </summary>
        /// <param name="capability"></param>
        /// <returns></returns>
        public bool SupportsCapability(string capability)
        {
            var supported = new[] { "vision", "tools", "json_mode" };

            return supported.Contains(capability.ToLower());
        }


        /// <summary>
        /// - Gets the AssembledItems list
        /// - Translates into OpenAI dialect
        /// </summary>
        /// <param name="concreteInvocation"></param>
        /// <returns> The input string to pass to the LLM </returns>
        /// <exception cref="Exception"></exception>
        private List<object> MappingAssembledInput(LlmInvocation concreteInvocation)
        {

            var inputItems = _assembler.AssembleInputItems(concreteInvocation);

            try
            {
                return inputItems.Select(x =>
                {
                    if (x.Type == "function_call")
                    {
                        return (object)new
                        {
                            type = x.Type,
                            name = x.ToolName,
                            arguments = JsonSerializer.Serialize(x.Arguments),
                            call_id = x.ToolCallId
                        };
                    }
                    else if (x.Type == "function_call_output")
                    {
                        return (object)new
                        {
                            type = x.Type,
                            call_id = x.ToolCallId,
                            output = x.Content
                        };
                    }
                    else
                    {
                        return (object)new
                        {
                            role = x.Role,
                            content = x.Content
                        };
                    }
                }).ToList();
            }
            catch (Exception ex)
            {
                throw new Exception($"[ERROR] OpenAIAdapter.MappingAssembledInput: {ex}");
            }

        }


        /// <summary>
        /// Extracts the model response string while avoiding other objects
        /// </summary>
        /// <param name="rawResponse"></param>
        /// <returns></returns>
        /// <exception cref="Exception"></exception>
        private string ExtractResponse(ModelResponse rawResponse)
        {
            string model_response = "";

            try
            {

                foreach (var item in rawResponse.Output)
                {
                    if (item.Type == "message")
                    {
                        foreach (var content in item.Content)
                        {
                            if (content.Type == "output_text")
                                model_response = content.Text;
                        }
                    }
                }
                return model_response;

            }
            catch (Exception ex)
            {
                throw new Exception($"[ERROR] The response received from the network call doesn't meet the correct parameters: {ex}");
            }
        }


        /// <summary>
        /// Extracts the model response in the tool usage domain, avoiding other objects
        /// OpenAI returns the arguments as objects -> deserializes them
        /// </summary>
        /// <param name="rawResponse"></param>
        /// <returns></returns>
        /// <exception cref="Exception"></exception>
        private List<ToolCallInvocation> ExtractToolInfo(ModelResponse rawResponse)
        {
            var toolList = new List<ToolCallInvocation>();

            try
            {
                foreach (var item in rawResponse.Output)
                {
                    if (item.Type == "function_call" || item.Type == "tool_call")
                    {
                        Dictionary<string, object> argsDict;

                        if (!string.IsNullOrEmpty(item.Arguments))
                            argsDict = JsonSerializer.Deserialize<Dictionary<string, object>>(item.Arguments);
                        else
                            argsDict = new Dictionary<string, object>();

                        var tool = new ToolCallInvocation
                        {
                            Type = "function_call",
                            CallId = item.Call_id,
                            ToolName = item.Name,
                            Arguments = argsDict
                        };

                        toolList.Add(tool);
                    }
                }


            }
            catch (Exception ex)
            {
                throw new Exception($"[ERROR] While parsing tool call: {ex}");
            }

            return toolList;

        }


        /// <summary>
        /// Builds tools schemas in OpenAI format
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns></returns>
        private List<OpenAIToolSchema> BuildToolSchema(ILlmInvocation invocation)
        {
            var toolSchemaList = new List<OpenAIToolSchema>();

            var builder = new OpenAIToolSchemaBuilder();

            if (invocation.Tools.RegisteredTools != null)
            {
                foreach (var tool in invocation.Tools.RegisteredTools)
                    toolSchemaList.Add(builder.Build(tool));
            }

            return toolSchemaList;
        }

    }
}
